polar-core/src/bindings.rs

```diff
@@ -425,52 +427,91 @@ impl BindingManager {
     }
 }
 
+fn combine(left: &Symbol, l: Operation, right: &Symbol, r: Operation) -> Operation {
+    let unis = l
+        .constraints()
+        .into_iter()
+        .chain(r.constraints().into_iter())
+        .filter_map(|op| match op {
+            Operation {
+                operator: Operator::Unify,
+                args,
+            } if args.len() == 2
+                && args[0].value().as_symbol().is_ok()
+                && args[1].value().as_symbol().is_ok() =>
+            {
+                Some((
+                    args[0].value().as_symbol().unwrap().clone(),
+                    args[1].value().as_symbol().unwrap().clone(),
+                ))
+            }
+            _ => None,
+        })
+        .chain(std::iter::once((left.clone(), right.clone())));
+    let sub = crate::data_filtering::partition_equivs(unis)
+        .into_iter()
+        .flat_map(|cls| {
+            if let Some(rep) = cls.iter().find(|v| !v.is_temporary_var()) {
+                cls.iter()
+                    .map(|pv| (pv.clone(), rep.clone()))
+                    .collect::<Vec<_>>()
+            } else {
+                cls.iter()
+                    .map(|pv| (pv.clone(), pv.clone()))
+                    .collect::<Vec<_>>()
+            }
+        })
+        .collect::<HashMap<_, _>>();
+
+    let subbed = |o| {
+        sub_vars(sub.clone(), term!(o))
+            .value()
+            .as_expression()
+            .unwrap()
+            .clone()
+    };
+    let base = Operation {
+        operator: Operator::And,
+        args: vec![],
+    };
+    base.merge_constraints(subbed(l))
+        .merge_constraints(subbed(r))
+}
+
 // Private impls.
 impl BindingManager {
     /// Bind two variables together.
     fn bind_variables(&mut self, left: &Symbol, right: &Symbol) -> Result<Option<Goal>> {
+        if left == right {
+            return Ok(None);
+        }
+        let pc = |cy: &Vec<Symbol>| cy.iter().map(|t| t.0.clone()).collect::<Vec<_>>();
+        use BindingManagerVariableState::*;
         let mut goal = None;
-        match (self._variable_state(left), self._variable_state(right)) {
-            (
-                BindingManagerVariableState::Bound(left_value),
-                BindingManagerVariableState::Unbound,
-            ) => {
-                self.add_binding(right, left_value);
-            }
-            (
-                BindingManagerVariableState::Unbound,
-                BindingManagerVariableState::Bound(right_value),
-            ) => {
-                self.add_binding(left, right_value);
+        match (
+            (left, self._variable_state(left)),
+            (right, self._variable_state(right)),
```

Comment:
The purpose of doing the match this way is to avoid having to write duplicate branches for left/right handed versions of the same case.